File: java/src/main/java/ai/smritea/sdk/model/SearchOptions.java
package ai.smritea.sdk.model;

/**
 * Builder-style options for searching memories.
 *
 * <p>Usage example:
 *
 * <pre>{@code
 * new SearchOptions()
 *     .withScope(MemoryScope.builder().actorId("alice").actorType("user").build())
 *     .withLimit(10);
 * }</pre>
 */
public final class SearchOptions {
  private MemoryScope scope;
  private Integer limit;
  private Float threshold;
  private Integer graphDepth;
  private String fromTime;
  private String toTime;
  private String validAt;

  /** Creates a new SearchOptions with all fields unset. */
  public SearchOptions() {}

  /**
   * Sets the memory scope (actor and conversation context).
   *
   * @param scope the memory scope
   * @return this instance for chaining
   */
  public SearchOptions withScope(MemoryScope scope) {
    this.scope = scope;
    return this;
  }

  /**
   * Sets the maximum number of results to return.
   *
   * @param limit the result limit
   * @return this instance for chaining
   */
  public SearchOptions withLimit(Integer limit) {
    this.limit = limit;
    return this;
  }

  /**
   * Sets the minimum relevance threshold for results.
   *
   * @param threshold the score threshold
   * @return this instance for chaining
   */
  public SearchOptions withThreshold(Float threshold) {
    this.threshold = threshold;
    return this;
  }

  /**
   * Sets the graph traversal depth for graph-aware search.
   *
   * @param graphDepth the graph depth
   * @return this instance for chaining
   */
  public SearchOptions withGraphDepth(Integer graphDepth) {
    this.graphDepth = graphDepth;
    return this;
  }

  /**
   * Sets the start of the time range filter (ISO 8601 format). Must be used together with toTime.
   *
   * @param fromTime the start of the time range (ISO 8601)
   * @return this instance for chaining
   */
  public SearchOptions withFromTime(String fromTime) {
    this.fromTime = fromTime;
    return this;
  }

  /**
   * Sets the end of the time range filter (ISO 8601 format). Must be used together with fromTime.
   *
   * @param toTime the end of the time range (ISO 8601)
   * @return this instance for chaining
   */
  public SearchOptions withToTime(String toTime) {
    this.toTime = toTime;
    return this;
  }

  /**
   * Sets a point-in-time filter (ISO 8601 format). Mutually exclusive with fromTime/toTime.
   *
   * @param validAt the point-in-time to filter by (ISO 8601)
   * @return this instance for chaining
   */
  public SearchOptions withValidAt(String validAt) {
    this.validAt = validAt;
    return this;
  }

  /** Returns the memory scope. */
  public MemoryScope getScope() {
    return scope;
  }

  /** Returns the result limit. */
  public Integer getLimit() {
    return limit;
  }

  /** Returns the score threshold. */
  public Float getThreshold() {
    return threshold;
  }

  /** Returns the graph depth. */
  public Integer getGraphDepth() {
    return graphDepth;
  }

  /** Returns the start of the time range filter. */
  public String getFromTime() {
    return fromTime;
  }

  /** Returns the end of the time range filter. */
  public String getToTime() {
    return toTime;
  }

  /** Returns the point-in-time filter. */
  public String getValidAt() {
    return validAt;
  }
}

File: csharp/SearchOptions.cs
// <copyright file="SearchOptions.cs" company="PlaceholderCompany">
// Copyright (c) PlaceholderCompany. All rights reserved.
// </copyright>

namespace Smritea.Sdk;

/// <summary>
/// Builder-style options for the SearchAsync method.
/// Use fluent <c>With*</c> methods for ergonomic construction:
/// <code>new SearchOptions().WithScope(new MemoryScope().WithActorId("alice")).WithLimit(10)</code>
/// </summary>
public sealed class SearchOptions
{
    /// <summary>Gets the scope containing actor and conversation context.</summary>
    public MemoryScope? Scope { get; private set; }

    /// <summary>Gets the maximum number of results to return.</summary>
    public int? Limit { get; private set; }

    /// <summary>Gets the minimum similarity threshold (0.0–1.0).</summary>
    public float? Threshold { get; private set; }

    /// <summary>Gets the graph traversal depth for graph-augmented search.</summary>
    public int? GraphDepth { get; private set; }

    /// <summary>Gets iSO-8601 datetime string — only return memories created at or after this time.</summary>
    public string? FromTime { get; private set; }

    /// <summary>Gets iSO-8601 datetime string — only return memories created at or before this time.</summary>
    public string? ToTime { get; private set; }

    /// <summary>Gets iSO-8601 datetime string — return memories valid at exactly this point in time.</summary>
    public string? ValidAt { get; private set; }

    /// <summary>Gets the search method override. Accepted values: "quick_search", "deep_search", "context_aware_search".</summary>
    public string? Method { get; private set; }

    /// <summary>Gets the reranker type override. Accepted values: "rrf_temporal", "rrf", "temporal", "node_distance", "mmr", "cross_encoder".</summary>
    public string? RerankerType { get; private set; }

    /// <summary>
    /// Gets the MongoDB-style operator DSL filter on memory metadata.
    /// Supports $eq, $ne, $gt, $gte, $lt, $lte, $in, $nin, $contains, $and, $or, $not, and wildcard "*".
    /// Values must be string, int, long, or double — booleans and nested objects are rejected by the server.
    /// Simple equality: <c>new Dictionary&lt;string, object&gt; { ["department"] = "engineering" }</c>
    /// Range: <c>new Dictionary&lt;string, object&gt; { ["level"] = new Dictionary&lt;string, object&gt; { ["$gte"] = 4 } }</c>
    /// Note: $contains is applied as a post-filter and may return fewer results than Limit.
    /// $contains inside $or is rejected with HTTP 400.
    /// </summary>
    public Dictionary<string, object>? MetadataFilter { get; private set; }

    /// <summary>Sets the scope containing actor and conversation context.</summary>
    /// <param name="scope">The scope object grouping actor and conversation fields.</param>
    /// <returns>The current instance for method chaining.</returns>
    public SearchOptions WithScope(MemoryScope scope)
    {
        this.Scope = scope;
        return this;
    }

    /// <summary>Sets the maximum number of results to return.</summary>
    /// <param name="limit">The maximum number of results (must be positive).</param>
    /// <returns>The current instance for method chaining.</returns>
    public SearchOptions WithLimit(int limit)
    {
        this.Limit = limit;
        return this;
    }

    /// <summary>Sets the minimum similarity threshold.</summary>
    /// <param name="threshold">The minimum score threshold in the range 0.0 to 1.0.</param>
    /// <returns>The current instance for method chaining.</returns>
    public SearchOptions WithThreshold(float threshold)
    {
        this.Threshold = threshold;
        return this;
    }

    /// <summary>Sets the graph traversal depth for graph-augmented search.</summary>
    /// <param name="graphDepth">The number of graph hops to traverse when expanding results.</param>
    /// <returns>The current instance for method chaining.</returns>
    public SearchOptions WithGraphDepth(int graphDepth)
    {
        this.GraphDepth = graphDepth;
        return this;
    }

    /// <summary>Sets the lower bound of the time range filter (ISO-8601).</summary>
    /// <param name="fromTime">ISO-8601 datetime string; only memories at or after this time are returned.</param>
    /// <returns>The current instance for method chaining.</returns>
    public SearchOptions WithFromTime(string fromTime)
    {
        this.FromTime = fromTime;
        return this;
    }

    /// <summary>Sets the upper bound of the time range filter (ISO-8601).</summary>
    /// <param name="toTime">ISO-8601 datetime string; only memories at or before this time are returned.</param>
    /// <returns>The current instance for method chaining.</returns>
    public SearchOptions WithToTime(string toTime)
    {
        this.ToTime = toTime;
        return this;
    }

    /// <summary>Sets a point-in-time filter (ISO-8601). Mutually exclusive with FromTime/ToTime.</summary>
    /// <param name="validAt">ISO-8601 datetime string; only memories valid at exactly this instant are returned.</param>
    /// <returns>The current instance for method chaining.</returns>
    public SearchOptions WithValidAt(string validAt)
    {
        this.ValidAt = validAt;
        return this;
    }

    /// <summary>Sets the search method override. Defaults to app config if omitted.</summary>
    /// <param name="method">Accepted values: "quick_search", "deep_search", "context_aware_search".</param>
    /// <returns>The current instance for method chaining.</returns>
    public SearchOptions WithMethod(string method)
    {
        this.Method = method;
        return this;
    }

    /// <summary>Sets the reranker type override. Only applies to deep_search. Defaults to app config if omitted.</summary>
    /// <param name="rerankerType">Accepted values: "rrf_temporal", "rrf", "temporal", "node_distance", "mmr", "cross_encoder".</param>
    /// <returns>The current instance for method chaining.</returns>
    public SearchOptions WithRerankerType(string rerankerType)
    {
        this.RerankerType = rerankerType;
        return this;
    }

    /// <summary>Sets the MongoDB-style operator DSL filter on memory metadata.</summary>
    /// <param name="metadataFilter">A dictionary using MongoDB-style operators ($eq, $ne, $gt, $gte, $lt, $lte, $in, $nin, $contains, $and, $or, $not, "*").</param>
    /// <returns>The current instance for method chaining.</returns>
    public SearchOptions WithMetadataFilter(Dictionary<string, object> metadataFilter)
    {
        this.MetadataFilter = metadataFilter;
        return this;
    }
}
